source/extensions/filters/http/rbac/rbac_filter.cc

```diff
@@ -23,28 +23,67 @@ RoleBasedAccessControlFilter::generateStats(const std::string& prefix, Stats::Sc
+RoleBasedAccessControlFilterConfig::engine(const Router::RouteConstSharedPtr route,
+                                           EnforcementMode mode) const {
   if (!route || !route->routeEntry()) {
-    return engine_;
+    return engine(mode);
   }
 
   const std::string& name = HttpFilterNames::get().RBAC;
   const auto* entry = route->routeEntry();
 
   const auto* route_local =
-      entry->perFilterConfigTyped<Filters::Common::RBAC::RoleBasedAccessControlEngine>(name)
+      entry->perFilterConfigTyped<RoleBasedAccessControlRouteSpecificFilterConfig>(name)
           ?: entry->virtualHost()
-                 .perFilterConfigTyped<Filters::Common::RBAC::RoleBasedAccessControlEngine>(name);
+                 .perFilterConfigTyped<RoleBasedAccessControlRouteSpecificFilterConfig>(name);
 
-  return route_local ? *route_local : engine_;
+  if (route_local) {
+    return route_local->engine(mode);
+  }
+
+  return engine(mode);
 }
 
+bool RoleBasedAccessControlFilterConfig::isEnabled(const Router::RouteConstSharedPtr route,
+                                                   EnforcementMode mode) const {
+  if (!route || !route->routeEntry()) {
+    return isEnabled(mode);
+  }
+
+  const std::string& name = HttpFilterNames::get().RBAC;
+  const auto* entry = route->routeEntry();
+
+  const auto* route_local =
+      entry->perFilterConfigTyped<RoleBasedAccessControlRouteSpecificFilterConfig>(name)
+          ?: entry->virtualHost()
+                 .perFilterConfigTyped<RoleBasedAccessControlRouteSpecificFilterConfig>(name);
+
+  if (route_local) {
+    return route_local->isEnabled(mode);
+  }
+
+  return isEnabled(mode);
+}
+
+RoleBasedAccessControlRouteSpecificFilterConfig::RoleBasedAccessControlRouteSpecificFilterConfig(
+    const envoy::config::filter::http::rbac::v2::RBACPerRoute& per_route_config)
+    : engine_(per_route_config.rbac().rules()),
+      shadow_engine_(per_route_config.rbac().shadow_rules()), config_(per_route_config.rbac()) {}
+
 Http::FilterHeadersStatus RoleBasedAccessControlFilter::decodeHeaders(Http::HeaderMap& headers,
                                                                       bool) {
-  const Filters::Common::RBAC::RoleBasedAccessControlEngine& engine =
-      config_->engine(callbacks_->route());
+  if (config_->isEnabled(callbacks_->route(), EnforcementMode::SHADOW)) {
+    if (config_->engine(callbacks_->route(), EnforcementMode::SHADOW)
+            .allowed(*callbacks_->connection(), headers)) {
+      config_->stats().shadow_allowed_.inc();
+    } else {
+      config_->stats().shadow_denied_.inc();
+    }
+  }
 
-  if (engine.allowed(*callbacks_->connection(), headers)) {
+  if (!config_->isEnabled(callbacks_->route(), EnforcementMode::ENFORCED) ||
+      config_->engine(callbacks_->route(), EnforcementMode::ENFORCED)
+          .allowed(*callbacks_->connection(), headers)) {
     config_->stats().allowed_.inc();
```

Comment:
Can we do not increase the stats when `config_->isEnabled()` is false? Otherwise I'm afraid the stats is not very useful as it is increased when rbac is not enabled.

Reply:
update the logic.